Allow to run without going through nvim and run busted directly

f4z3r opened the pull request:
Relates to: #58

I have not implemented a check for PATH as I am not 100% sure that it is really necessary. You might want to configure this to run `busted` from your PATH but only have it configured in some repositories using some form of shimming. Then it would not be very nice if you get a notification that it cannot find `busted` every time you open a non-Lua project where you do not have `busted` set up.

MisanthropicBit commented on lua/neotest-busted/init.lua:
The `no_nvim` option is missing from the directory-local case.

MisanthropicBit commented on lua/neotest-busted/init.lua:
```diff
@@ -191,18 +194,9 @@ local function quote_string(string)
-
+---@param busted_config neotest-busted.BustedCommandConfig
+---@return arguments string[]
+local function get_nvim_arguments(busted_config)
```
The sections adding `config.busted_paths`, `config.busted_cpaths`, `busted_config.lua_paths`, and `busted_config.lua_cpaths` have to be added in the general case, not just for neovim, as someone running `busted` directly might want to add custom paths and we still want to include the paths we find via `find_busted_command`.

```lua
    -- Add paths for busted
    vim.list_extend(lua_paths, busted_config.lua_paths)
    vim.list_extend(lua_cpaths, busted_config.lua_cpaths)
```

In relation to the above, please add tests in `adapter_build_spec_spec.lua` so we are testing that the right commands are being constructed and we don't break existing functionality.

f4z3r replied:
Makes sense, I updated the code to inject the paths via the environment in case `busted` is run without nvim.

MisanthropicBit commented on lua/neotest-busted/init.lua:
```diff
@@ -307,7 +315,7 @@ function BustedNeotestAdapter.create_test_command(paths, options)
 
     return {
-        nvim_command = compat.loop.exepath(),
+        nvim_command = config.no_nvim and busted_command or compat.loop.exepath(),
```
I think we should change `nvim_command` to just `command` as it isn't necessarily a neovim command anymore.

MisanthropicBit reviewed (changes requested):
Thanks for your hard work 🙂 I am still be a unsure how useful this addition is but adding it should be fine.

> I have not implemented a check for PATH as I am not 100% sure that it is really necessary

I agree. Relying on `PATH` is also not always very deterministic as you rightly explain. This can added if someone has a good use case.

f4z3r commented:
> I am still be a unsure how useful this addition is but adding it should be fine.

Fair enough. As I said in the linked issue, I can get it to work some other way. If you think it is not worth it, no hard feelings if you just close the PR. It was fun to look into how neotest adapters are written either way ☺️

cshuaimin commented:
I can give you an example of how I use it. I'm working on an OpenResty project, but the build environment doesn't work well with my OS. I can only build it using Docker containers. To run its tests, busted also needs to be called in the container. I can't use nvim as the Lua interpreter to run busted – it's either hard or not possible.
I have written a shell script to wrap busted so I don't need to exec into the container manually:

```console
bat -p /nix/store/50ww9mz6jjacyipb3a4ifl81igv8fq73-busted/bin/busted
#!/nix/store/q7sqwn7i6w2b67adw0bmix29pxg85x3w-bash-5.3p3/bin/bash
exec distrobox enter dev --no-tty -- bin/busted "$@"
```

I tested this branch and it works very well for me.

f4z3r commented:
@cshuaimin Ah, a fellow NixOS user 😍 I am a bit surprised, that the paths are reset. I will have a look at this today. Whether to use busted directly by default if nothing is found would change the behaviour of existing configs. I guess @MisanthropicBit should say what he thinks about this.

cshuaimin commented:
@f4z3r Oh I see, then that issue 1 is OK. (Loving NixOS btw.)

cshuaimin commented:
@f4z3r You are awesome! I can confirm it works!

MisanthropicBit commented:
> Fair enough. As I said in the linked issue, I can get it to work some other way. If you think it is not worth it, no hard feelings if you just close the PR. It was fun to look into how neotest adapters are written either way ☺️

That's very nice of you, thanks. Although it seems @cshuaimin has been kind enough to chime in and provide a legitimate reason. Another benefit is that the `busted_command` option doesn't need to be a cli lua runner script anymore but can be the actual busted executable (which is a shell script) or a custom shell script like shown above since it doesn't need to run through neovim.

> I have to set busted_command = "busted", it cannot find busted without this line. I think maybe it should default to busted?

I'm a bit against this because I generally don't like falling back to `PATH` as it can change unexpectedly. By default `neotest-bsuted` will only look for a project-local installation of busted since it can be installed in three different locations with different versions, although to be fair the last version of busted got released 2 years ago.

In the case where `busted_command` is set, `path` and `cpath` are set to empty lists because we cannot know anything about where the paths actually are. It is expected that the user provides them.

> alright, do you want to give it a try? I changed the code so that the configured busted_paths and busted_cpaths are appended to the ones set in the env when not using nvim.

I don't think we should have this special-case state dependency based on the value of `no_nvim` if I understand correctly. That will be confusing for users I think. Alternatively:

1. If setting `busted_command` to the busted executable (shell script, not cli runner), package paths are already statically set in the shell script.
2. The way other adapters handle dynamic options is to allow the option to accept a function that takes as argument the default value, if any, and the user can return whatever they want. The default function would simply return its argument. In this way, each user can have exact control for their specific scenario. See [here]() for an example.
3. Allowing users to set a custom environment via a config option. Also another common adapter strategy.

See [here](https://github.com/nvim-neotest/neotest-jest/blob/main/lua/neotest-jest/init.lua#L552-L584) for an example of setting config options via functions or a value.

@cshuaimin @f4z3r I'm not a nixos user (although I get the feeling that I should be) but would any of the above suit your specific scenarios?

f4z3r commented on lua/neotest-busted/init.lua:
```diff
@@ -568,15 +595,24 @@ function BustedNeotestAdapter.build_spec(args)
         vim.list_extend(command, args.extra_args)
     end
 
+    local env = nil
+    if test_command.set_env then
+        env = {
+            LUA_PATH = util.normalize_and_create_lua_path(unpack(test_command.paths)),
+            LUA_CPATH = util.normalize_and_create_lua_path(unpack(test_command.cpaths)),
+        }
+    end
```
Not necessarily, they are complemented with the configuration in the configuration based on what people set in `busted_paths` and `busted_cpaths`.

cshuaimin commented on lua/neotest-busted/init.lua:
```diff
@@ -568,15 +588,24 @@ function BustedNeotestAdapter.build_spec(args)
     end
 
+    local env = nil
+    if test_command.set_env then
+        env = {
+            LUA_PATH = util.normalize_and_create_lua_path(unpack(test_command.paths)),
+            LUA_CPATH = util.normalize_and_create_lua_path(unpack(test_command.cpaths)),
+        }
+    end
```
I'm OK with both ways now: append or override LUA_PATH (even not providing a config). I moved the LUA_PATH setup into the container, so LUA_PATH is always empty in nvim. You may choose what you think is simpler. Thanks!

f4z3r commented:
> Sorry for all the comments 😅 I appreciate the work you're doing on the PR!

No worries, glad for the feedback. I added your suggestions and changed how functions in the config are evaluated when called on the config. I decided to prepend the configured paths to the environment, as I would expect the paths set in the `neotest-busted` configuration to take precedence over the environment.
